server/src/main/java/org/elasticsearch/index/mapper/SyntheticSourceDocumentParserListener.java

```diff
@@ -0,0 +1,192 @@
+                    data.endObject();
+
+                    if (startingToken instanceof Token.StartObject) {
+                        depth -= 1;
+                        if (depth == 0) {
+                            var parentOffset = parentMapper.fullPath().length() + 1;
+                            // TODO the way we store values is not final, maybe we should put them directly in DocumentParserContext ?
+                            // does not feel great though
+                            valuesToStore.add(
+                                new IgnoredSourceFieldMapper.NameValue(fullPath, parentOffset, XContentDataHelper.encodeXContentBuilder(data), document)
+                            );
+                            return returnState;
+                        }
+
+                    }
+                }
+                case Token.StartArray ignored -> {
+                    data.startArray();
+                    if (startingToken instanceof Token.StartArray) {
+                        depth += 1;
+                    }
+                }
+                case Token.EndArray ignored -> {
+                    data.endArray();
+
+                    if (startingToken instanceof Token.StartArray) {
+                        // TODO extract function
+                        depth -= 1;
+                        if (depth == 0) {
+                            var parentOffset = parentMapper.fullPath().length() + 1;
+                            // does not feel great though
+                            valuesToStore.add(
+                                new IgnoredSourceFieldMapper.NameValue(fullPath, parentOffset, XContentDataHelper.encodeXContentBuilder(data), document)
+                            );
+                            return returnState;
+                        }
+                    }
+                }
+                case Token.FieldName fieldName -> data.field(fieldName.name());
+                case Token.StringValue stringValue -> data.value(stringValue.value());
+                case Token.BooleanValue booleanValue -> data.value(booleanValue.value());
+                case Token.IntValue intValue -> data.value(intValue.value());
+            }
+
+            return this;
+        }
+    }
+
+    class Watching implements State {
+        private String fullPath;
+        private Stack<ObjectMapper> parents;
+        private Stack<String> objectArrays;
+        private Mapper currentMapper;
+
+        Watching(String fullPath, Stack<ObjectMapper> parents, Mapper currentMapper) {
+            this.fullPath = fullPath;
+            this.parents = parents;
+            this.objectArrays = new Stack<>();
+            this.currentMapper = currentMapper;
+        }
+
+        public State consume(Token token) throws IOException {
+            switch (token) {
+                case Token.StartObject startObject -> {
+                    if (currentMapper instanceof ObjectMapper om && om.isEnabled() == false) {
```

Comment:
So we'd need similar checks for all the cases where we want to use ignored source, right? These require accessing tha parent object mapper and the field mapper, so they should apply. In some cases, we also use the parsing context I think.. It may help to pass it to `consume` as well.

Reply:
I don't want to have one shared context between multiple listeners because i think it would be hard to reason about. Unless it's immutable which is an option. I kind of lean towards each listener establishing whatever context it needs on its own.

Reply:
What about the other cases where we need ignored source?